services/api/service.go

```diff
@@ -950,13 +1097,35 @@ func (api *RelayAPI) handleSubmitNewBlock(w http.ResponseWriter, req *http.Reque
 		}
 	}
 
-	builderIsHighPrio, builderIsBlacklisted, err := api.redis.GetBlockBuilderStatus(payload.Message.BuilderPubkey.String())
+	builderStatus, err := api.redis.GetBlockBuilderStatus(payload.Message.BuilderPubkey.String())
```

Comment:
We should maybe consider caching `api.redis.GetBlockBuilderStatus` and `api.redis.GetBlockBuilderCollateral` for all builders, e.g. taking a snapshot once per slot. Here's my thinking:

1) With optimistic relaying we can have looser rate limiting (e.g. 100 bids/sec/builder). Each Redis read takes on the order of 1ms so these Redis calls can add up, even with just tens of builder pubkeys. The current codebase was designed for a 2 bids/sec/builder paradigm and may need a bit of performance refactoring to handle the load.
2) The status of builders can change mid-slot in two different ways: a) through demotions, and b) through the internal API. As such there are potential race conditions whenever the internal API is used. It would be conceptually cleaner if at the beginning of every slot (e.g. in `updateOptimisticSlot`) we take a snapshot of the relevant subset of `vars.TableBlockBuilder`.
3) Those Redis calls are in the critical path so we're looking to shave a couple milliseconds of latency—a nice-to-have micro-optimisation.

Reply:
i think this is a nice change! i added logic to updateOptimisticSlot to cache (create a map as a field of the relay type) the builder status and collaterals at each slot by reading from the TableBlockBuilder as you mentioned. then in the critical path of the block submission, we can just read from this cache. this has a small added benefit of getting rid of our need for the Get and Set collateral on the redis interface (because we only ever read the collateral from the db). and it makes the code more readable on the fast path --- feels like a win! 

i was thinking about if we should filter the builders put in the cache, but I don't really think its necessary. we could use a heuristic like "only add builders who have submitted a block in the last 100 slots or something", but unless we need to this feels like http://wiki.c2.com/?PrematureOptimization :-) especially considering this is happening at the beginning of each slot, one db read and iterating through the list of builders just doesn't seem that crazy to me. but lmk what you think!

Reply:
> lmk what you think!

I think it looks great :)